src/compiler/utils/string_with_sourcemap.ts

```diff
@@ -0,0 +1,269 @@
+		return new StringWithSourcemap(string, map);
+	}
+
+	static from_source(
+		source_file: string, source: string, offset?: SourceLocation
+	): StringWithSourcemap {
+		if (!offset) offset = { line: 0, column: 0 };
+		const map: DecodedSourceMap = { version: 3, names: [], sources: [source_file], mappings: [] };
+		if (source == '') return new StringWithSourcemap(source, map);
+
+		// we create a high resolution identity map here,
+		// we know that it will eventually be merged with svelte's map,
+		// at which stage the resolution will decrease.
+		const line_list = source.split('\n');
+		for (let line = 0; line < line_list.length; line++) {
+			map.mappings.push([]);
+			const token_list = line_list[line].split(/([^\d\w\s]|\s+)/g);
+			for (let token = 0, column = 0; token < token_list.length; token++) {
+				if (token_list[token] == '') continue;
+				map.mappings[line].push([column, 0, offset.line + line, column]);
+				column += token_list[token].length;
+			}
+		}
+		// shift columns in first line
+		const segment_list = map.mappings[0];
+		for (let segment = 0; segment < segment_list.length; segment++) {
+			segment_list[segment][3] += offset.column;
+		}
+		return new StringWithSourcemap(source, map);
+	}
+}
+
+export function combine_sourcemaps(
+	filename: string,
+	sourcemap_list: Array<DecodedSourceMap | RawSourceMap>
+): RawSourceMap {
+	if (sourcemap_list.length == 0) return null;
+
+	let map_idx = 1;
+	const map: RawSourceMap =
+		sourcemap_list.slice(0, -1)
+			.find(m => m.sources.length !== 1) === undefined
+
+			? remapping( // use array interface
+				// only the oldest sourcemap can have multiple sources
+				sourcemap_list,
+				() => null,
+				true // skip optional field `sourcesContent`
+			)
+
+			: remapping( // use loader interface
+				sourcemap_list[0], // last map
+				function loader(sourcefile) {
+					if (sourcefile === filename && sourcemap_list[map_idx]) {
+						return sourcemap_list[map_idx++]; // idx 1, 2, ...
+						// bundle file = branch node
+					}
+					else return null; // source file = leaf node
+				} as SourceMapLoader,
+				true
+			);
+
+	if (!map.file) delete map.file; // skip optional field `file`
+
+	return map;
+}
+
+// browser vs node.js
+const b64enc = typeof btoa == 'function' ? btoa : b => Buffer.from(b).toString('base64');
+
+export function apply_preprocessor_sourcemap(filename: string, svelte_map: SourceMap, preprocessor_map_input: string | DecodedSourceMap | RawSourceMap): SourceMap {
+	if (!svelte_map || !preprocessor_map_input) return svelte_map;
+
+	const preprocessor_map = typeof preprocessor_map_input === 'string' ? JSON.parse(preprocessor_map_input) : preprocessor_map_input;
+
+	const result_map = combine_sourcemaps(
+		filename,
+		[
+			svelte_map as RawSourceMap,
+			preprocessor_map
+		]
+	) as RawSourceMap;
+
+	//Svelte expects a SourceMap which includes toUrl and toString. Instead of using the magic-string constructor that takes a decoded map
```

Comment:
we usually have a space after `//`. It'd probably make sense to do that here and on the next line for consistency